File: src/main/algorithms/Connections.cpp
/* ---------------------------------------------------------------------
 * Numenta Platform for Intelligent Computing (NuPIC)
 * Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
 * with Numenta, Inc., for a separate license for this software code, the
 * following terms and conditions apply:
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 *
 * http://numenta.org/licenses/
 * ----------------------------------------------------------------------
 */

/** @file
 * Implementation of Connections
 */

#include <iostream>
#include <nta/algorithms/Connections.hpp>

using namespace std;
using namespace nta;
using namespace nta::algorithms::connections;

Connections::Connections(CellIdx numCells) : cells_(numCells)
{
}

Segment Connections::createSegment(const Cell& cell)
{
  vector<SegmentData>& segments = cells_[cell.idx].segments;
  Segment segment = {segments.size(), cell};

  SegmentData segmentData;
  segments.push_back(segmentData);

  return segment;
}

Synapse Connections::createSynapse(const Segment& segment,
                                   const Cell& presynapticCell,
                                   Permanence permanence)
{
  const Cell& cell = segment.cell;

  vector<SynapseData>& synapses = cells_[cell.idx].segments[segment.idx].synapses;
  Synapse synapse = {synapses.size(), segment};

  SynapseData synapseData = {presynapticCell, permanence};
  synapses.push_back(synapseData);

  return synapse;
}

void Connections::updateSynapsePermanence(const Synapse& synapse,
                                          Permanence permanence)
{
  const Segment& segment = synapse.segment;
  const Cell& cell = segment.cell;

  cells_[cell.idx].segments[segment.idx].synapses[synapse.idx].permanence = permanence;
}

vector<Segment> Connections::getSegmentsForCell(const Cell& cell)
{
  vector<Segment> segments;
  Segment segment;

  for(SegmentIdx i = 0; i < cells_[cell.idx].segments.size(); i++) {
    segment.idx = i;
    segment.cell = cell;
    segments.push_back(segment);
  }

  return segments;
}

vector<Synapse> Connections::getSynapsesForSegment(const Segment& segment)
{
  const Cell& cell = segment.cell;
  vector<Synapse> synapses;
  Synapse synapse;

  for(SynapseIdx i = 0; i < cells_[cell.idx].segments[segment.idx].synapses.size(); i++) {
    synapse.idx = i;
    synapse.segment = segment;
    synapses.push_back(synapse);
  }

  return synapses;
}

SynapseData Connections::getDataForSynapse(const Synapse& synapse) const
{
  const Segment& segment = synapse.segment;
  const Cell& cell = segment.cell;

  return cells_[cell.idx].segments[segment.idx].synapses[synapse.idx];
}

bool Connections::getMostActiveSegmentForCells(const std::vector<Cell>& cells,
                                               const std::vector<Cell>& input,
                                               UInt synapseThreshold,
                                               Segment& segment) const
{
  // UInt numSynapses, maxSynapses = synapseThreshold;
  // vector<SegmentData> segments;
  // vector<SynapseData> synapses;
  bool found = false;

  // for (vector<Cell>::const_iterator c = cells.begin(); c != cells.end(); c++) {
  //   segments = cells_[c->idx].segments;

  //   for (vector<SegmentData>::const_iterator s = segments.begin(); s != segments.end(); s++) {
  //     synapses = s->synapses;
  //     numSynapses = 0;

  //     for (vector<SynapseData>::const_iterator syn = synapses.begin(); syn != synapses.end(); syn++) {
  //       if (find(input.begin(), input.end(), syn->presynapticCell) != input.end()) {  // TODO: Optimize this
  //         numSynapses++;
  //       }
  //     }

  //     if (numSynapses >= maxSynapses) {
  //       maxSynapses = numSynapses;
  //       segment.idx = s - segments.begin();
  //       segment.cell = *c;
  //       found = true;
  //     }
  //   }
  // }

  return found;
}

Activity Connections::computeActivity(const std::vector<Cell>& input,
                                      Permanence permanenceThreshold,
                                      UInt synapseThreshold) const
{
  Activity activity;
  // vector<Synapse> synapses;
  // SynapseData synapseData;
  // Synapse synapse;

  // for (vector<Cell>::const_iterator c = input.begin(); c != input.end(); c++) {
  //   synapses = synapsesForPresynapticCell_[*c];

  //   for (vector<Synapse>::const_iterator s = synapses.begin(); s != synapses.end(); s++) {
  //     synapseData = getDataForSynapse(*s);

  //     if (synapseData.permanence >= permanenceThreshold) {
  //       activity.numActiveSynapsesForSegment[synapse.segment] += 1;
  //     }

  //     if (activity.numActiveSynapsesForSegment[synapse.segment]) {
  //       activity.numActiveSegmentsForCell[*c] += 1;
  //     }
  //   }
  // }

  return activity;
}

File: py/htm/advanced/algorithms/connections.py
# ------------------------------------------------------------------------------
# HTM Community Edition of NuPIC
# Copyright (C) 2019, Frederick C. Rotbart
#
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero Public License version 3 as published by the Free
# Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero Public License for more details.
#
# You should have received a copy of the GNU Affero Public License along with
# this program.  If not, see http://www.gnu.org/licenses.
# ------------------------------------------------------------------------------

from htm.bindings.algorithms import Connections as CPPConnections
import numpy as np

class Connections(CPPConnections):
    
    def numConnectedSynapsesForCells(self, cells):
        """
        Return the number of connected synapses in the connection for the list of cells.
        """
        n = 0
        for cell in cells:
            segments = self.segmentsForCell(cell)
            for segment in segments:
                n += self.numConnectedSynapses(segment)
        return n
    
    def numSynapsesForCells(self, cells):
        """
        Return the number of connected synapses in the connection for the list of cells.
        """
        n = 0
        for cell in cells:
            segments = self.segmentsForCell(cell)
            for segment in segments:
                n += self.numSynapses(segment)
        return n
    
    def numSegmentsWithSynapses(self, cells):
        """
        Return the number of segments in the connection that have at least one synapse for the list of cells.
        """
        n = 0
        for cell in cells:
            segments = self.segmentsForCell(cell)
            for segment in segments:
                if self.numSynapses(segment) > 0:
                    n += 1
        return n
    
    def sortSegmentsByCell(self, segments):
        """
        Sort an array of segments by cell in increasing order.
        
        @param segments
            The segment array.
            
        @return:
            A sorted segment array
        """
        cells = [self.cellForSegment(s) for s in segments]
        cells_args = np.argsort(cells)
        return segments[cells_args]
    
    def filterSegmentsByCell(self, segments, cells, assumeSorted=False):
        """
        Return the subset of segments that are on the provided cells.
        
        @param segments
            The segments to filter. Must be sorted by cell.
        
        @param cells
            The cells whose segments we want to keep. Must be sorted.
        
        """
        if not assumeSorted:
            segments = self.sortSegmentsByCell(segments)

        mask = np.isin([self.cellForSegment(s) for s in segments], cells)
        return segments[mask]   

    def mapSegmentsToCells(self, segments):
        """
        Get the cell for each provided segment.
        
        @param segments
            The segments to query
        
        @param cells
            Output array with the same length as 'segments'
        """
        return np.array([self.cellForSegment(s) for s in segments], dtype=np.uint32)

    def growSynapsesToSample(self, segment, growthCandidates, maxNew, initialPermanence, rng):
        """
        For each specified segments, grow synapses to a random subset of the
        inputs that aren't already connected to the segment.
        *
        @param segment
        The segment to modify
        *
        @param inputs
        The inputs to sample
        *
        @param sampleSize
        The number of synapses to attempt to grow per segment
        *
        @param initialPermanence
        The permanence for each added synapse
        *
        @param rng
        Random number generator
        """
        presynamptic_cells = [self.presynapticCellForSynapse(synapse) for synapse in self.synapsesForSegment(segment)]
        active_cells_without_synapses = growthCandidates[np.isin(growthCandidates, presynamptic_cells, invert=True)]
        if len(active_cells_without_synapses) > maxNew:
            active_cells_without_synapses = rng.sample(np.asarray(active_cells_without_synapses, dtype="uint32"), maxNew)
        
        for c in active_cells_without_synapses:
            self.createSynapse(segment, c, initialPermanence)
    
    def getSegmentCounts(self, cells):
        """
        Get the number of segments on each of the provided cells.
        
        @param cells
        The cells to check
        
        @param counts
        Output array with the same length as 'cells'
        """
        return np.array([self.numSegments(cell) for cell in cells], dtype=np.uint32)
        
    def computeActiveSegments(self, presynapticCells, activationThreshold):
        """
        Compute the segments whose number of active synapses is greater or equal to activationThreshold
        for a vector of active presynaptic cells
        
        @param SDR presynapticCells
        The cells to check
        
        @param int activationThreshold
        The threshold that number of synapses must reach
        
        @return list
        List of segments with greater or equal number of sysnapses to activationThreshold 
        """
        overlaps = self.computeActivity(presynapticCells, False)
        return np.flatnonzero(overlaps >= activationThreshold)

